### PR TITLE
Feature flexible editor types

### LINKED ISSUE #398: Multiple choice exercise type
The teacher should be able to decide whether the student has to pick one or more answers.

### LINKED ISSUE #400: Editor height proportional to expected solution size
We should add an exercise parameter that states the expected solution size. Possible values could be:
- `one_liner` - a single line
- `small` - 5 lines
- `medium` - 15 lines
- `large`  - 50 lines

### LINKED ISSUE #600: If you upload a solution, you don't know which is the code currently submitted

### LINKED ISSUE #661: Tabs should not being displayed if there is only and editor or a console in exercise view

### PR DESCRIPTION
**Pull request by flbulgarelli**
Fixes #661
Fixes #398, since this PR adds single and multiple choices exercises. 
Fixes #400, since this PR allows to have one liners and large code editor. 
Fixes #600, since it adds a download link to get your solution as a file with the appropriate extension

It also adds placeholders to code editor, and removes the print margin - the annoying grey line - from it, and renamed the editor -> solution tab. 

Depends on https://github.com/mumuki/mumuki-teacher-tools/issues/186

### CODE REVIEW

**Comment by flbulgarelli** [comment]
@faloi with this new feature, we are actually adding new exercise modelling alternatives, since the editor types are available to all runners, not only text-runner:

* It is easier and more evident to request one-liners as a solution, using the new text-input editor
* you can do queries even if you are using the upload mode, so you can request the student to try things in the console, and then upload a solution
* you can do choice exercises asking for a specific text response, or multiple responses
* it is easier to add block languages support
* you can ask the user which piece of code will work. And you can even let the user to test each one of them, being an intermediate point between a playground or copy/past exercise, and full code-from-scratch exercise. For example:

```
> Veamos si se entiende: ¿cual de las siguientes tres expresiones funcionará? 
>  * (const undefined) 5  
>  * (const 5) undefined
> * undefined
```

**Comment by flbulgarelli** [comment]
Pending tasks:

* Properly render multiple choices
* Show uploaded content

**Comment by flbulgarelli** [comment]
Choice: 

screencapture-central-localmumuki-io-3000-exercises-20-1476823034907

**Comment by flbulgarelli** [comment]
Multiple choice: 

screencapture-central-localmumuki-io-3000-exercises-20-1476885702477

**Comment by flbulgarelli** [comment]
Code editor - today'd editor.

screencapture-central-localmumuki-io-3000-exercises-20-1476885813369

**Comment by flbulgarelli** [comment]
Upload, with `input_bottom` layout -  now it supports console, too.

screencapture-central-localmumuki-io-3000-exercises-20-1476886065715

**Comment by flbulgarelli** [comment]
Single line - `text` editor: 

screencapture-central-localmumuki-io-3000-exercises-68-programacion-funcional-valores-y-funciones-mas-composicion-1476889719591

File: app/assets/stylesheets/application/modules/_exercise_assignment.scss [hunk 1]
```diff
@@ -6,7 +6,28 @@
...
+.mu-editor-overlap {
   z-index: 0;
   position: relative;
 }
+
+.mu-editor {
+  .form-group {
+    margin: 10px 0;
+  }
+  input[type=radio] {
```
Comment (aguspina) [inline app/assets/stylesheets/application/modules/_exercise_assignment.scss]: @flbulgarelli shall we add more styles to the checkboxes and radio buttons? like complementary color for the check?

File: app/models/exercise.rb [hunk 2]
```diff
@@ -84,7 +84,7 @@ def new_solution
   end
 
   def extra_preview
-    "```#{language.name}\n#{extra}```"
+    Mumukit::ContentType::Markdown.highlighted_code(language.name, extra)
```
Comment (aguspina) [inline app/models/exercise.rb]: Nice

**Comment by aguspina** [comment]
@flbulgarelli This is a nice PR!

File: app/assets/stylesheets/application/modules/_exercise_assignment.scss (hunk 1 above)
Comment (flbulgarelli) [inline app/assets/stylesheets/application/modules/_exercise_assignment.scss]: Yes, I think so. However I prefer to first merge this PR and later improve its styling. Would you want to open an issue and submit proposal about how it should look?

**Pull request merged**